### PR TITLE
Kubelet: Cache image history to eliminate the performance regression

### PR DESCRIPTION
**Pull request by Random-Liu**
Fix https://github.com/kubernetes/kubernetes/issues/25057.

The image history operation takes almost 50% of cpu usage in kubelet performance test. We should cache image history instead of getting it from runtime everytime.

This PR cached image history in imageStatsProvider and added unit test.

@yujuhong @vishh 
/cc @kubernetes/sig-node 

Mark v1.3 because this is a relatively significant performance regression.

[Analytics]()

### CODE REVIEW

File: pkg/kubelet/dockertools/images.go [hunk 1]
```diff
@@ -18,54 +18,85 @@ package dockertools
...
 	images, err := isp.c.ListImages(dockertypes.ImageListOptions{})
 	if err != nil {
 		return nil, fmt.Errorf("failed to list docker images - %v", err)
 	}
-	// A map of all the image layers to its corresponding size.
-	imageMap := sets.NewString()
-	ret := &runtime.ImageStats{}
+	// Create new cache each time, this is a little more memory consuming, but:
+	// * ImageStats is only called every 10 seconds
+	// * There shouldn't be too many images
```
Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: I don't think this is a valid assumption. The added memory cost might be worth the performance overhead though.

File: pkg/kubelet/dockertools/images.go [hunk 2]
```diff
@@ -18,54 +18,85 @@ package dockertools
...
+func newImageStatsProvider(c DockerInterface) *imageStatsProvider {
+	return &imageStatsProvider{
+		layers:          make(map[string]*dockertypes.ImageHistory),
+		imageToLayerIDs: make(map[string][]string),
+		c:               c,
+	}
+}
+
 func (isp *imageStatsProvider) ImageStats() (*runtime.ImageStats, error) {
+	isp.Lock()
```
Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: nit: Why not take the lock after listing images?

File: pkg/kubelet/dockertools/images.go [hunk 3]
```diff
@@ -18,54 +18,85 @@ package dockertools
...
-			continue
-		}
-		// Store size information of each layer.
-		for _, layer := range history {
-			// Skip empty layers.
-			if layer.Size == 0 {
-				glog.V(10).Infof("skipping image layer %v with size 0", layer)
+		layerIDs, ok := isp.imageToLayerIDs[image.ID]
+		if !ok {
+			// Get information about the various layers of each docker image.
```
Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: For each docker image or for a given image?

File: pkg/kubelet/dockertools/images.go [hunk 4]
```diff
@@ -18,54 +18,85 @@ package dockertools
...
 	"github.com/golang/glog"
 
 	dockertypes "github.com/docker/engine-api/types"
 	runtime "k8s.io/kubernetes/pkg/kubelet/container"
-	"k8s.io/kubernetes/pkg/util/sets"
 )
 
 // imageStatsProvider exposes stats about all images currently available.
 type imageStatsProvider struct {
+	sync.Mutex
```
Comment (vishh) [inline pkg/kubelet/dockertools/images.go]: Why is this lock needed?

**Comment by vishh** [comment]
Thanks for the PR @Random-Liu! Just a few comments.

File: pkg/kubelet/dockertools/images.go (hunk 1 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/images.go]: ACK. Will remove this one.

File: pkg/kubelet/dockertools/images.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/images.go]: ACK.

File: pkg/kubelet/dockertools/images.go (hunk 3 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/images.go]: Given one. I just copied the comment here, will update it.

File: pkg/kubelet/dockertools/images.go (hunk 4 above)
Comment (Random-Liu) [inline pkg/kubelet/dockertools/images.go]: Discussed with @timstclair, and he said it is possible that summary api is called by multiple threads, so I added the lock here. :)

**Comment by Random-Liu** [comment]
@k8s-bot test this issue #26345.

**Comment by Random-Liu** [comment]
@vishh Addressed comments and thanks for reviewing! :)

**Pull request merged**